### PR TITLE
Pyethereum keystore support

### PR DESCRIPTION
**Pull request by tworec**
This will enable to read keystores created via https://github.com/ethereum/pyethereum which we are using in https://github.com/golemfactory/golem.

This PR will allow to read only one such a keystore because there is hardcoded 0-address imputed, and address is a key in internal map of accounts. 

see https://github.com/ethereum/pyethereum/issues/739

### CODE REVIEW

File: ethstore/src/account/crypto.rs [hunk 1]
```diff
@@ -79,12 +79,14 @@ impl Crypto {
 
 	/// Encrypt custom plain data
 	pub fn with_plain(plain: &[u8], password: &Password, iterations: u32) -> Result<Self, crypto::Error> {
-		let salt: [u8; 32] = Random::random();
+		let salt_arr: [u8; 32] = Random::random();
```
[the remark this answers is not in the document]
Comment (niklasad1) [reply ethstore/src/account/crypto.rs]: Yeah, I wonder the same since [crypto::derive_key_iterations](https://github.com/paritytech/parity-common/blob/master/parity-crypto/src/lib.rs#L58) requires a slice and it should be fine to allocate an array on the stack and pass a reference to it!

File: ethstore/cli/src/main.rs [hunk 2]
```diff
@@ -145,7 +145,6 @@ impl fmt::Display for Error {
 }
 
 fn main() {
-	panic_hook::set_abort();
```
Comment (niklasad1) [inline ethstore/cli/src/main.rs]: Why is this removed?

Comment (debris) [reply ethstore/cli/src/main.rs]: same question

File: Cargo.lock [hunk 3]
```diff
@@ -1965,11 +1964,11 @@ dependencies = [
 [[package]]
 name = "parity-crypto"
 version = "0.1.0"
-source = "git+https://github.com/paritytech/parity-common#0045887fecd2fec39e56c962a0b27e2caf3b9474"
+source = "git+https://github.com/paritytech/parity-common#51258799ee639783a7153d807ce9ddf4d49b29fa"
 dependencies = [
  "ethereum-types 0.3.2 (registry+https://github.com/rust-lang/crates.io-index)",
  "quick-error 1.2.2 (registry+https://github.com/rust-lang/crates.io-index)",
- "ring 0.12.1 (git+https://github.com/paritytech/ring)",
+ "ring 0.12.0 (registry+https://github.com/rust-lang/crates.io-index)",
```
Comment (debris) [inline Cargo.lock]: what's the reason to change `ring` dependency? iirc the official repo does not contain several changes that we need

File: Cargo.lock [hunk 4]
```diff
@@ -3684,6 +3679,11 @@ dependencies = [
  "xml-rs 0.7.0 (registry+https://github.com/rust-lang/crates.io-index)",
 ]
 
+[[patch.unused]]
+name = "ring"
+version = "0.12.1"
+source = "git+https://github.com/paritytech/ring#bae475e9f7ea7dd4ae671bef4b576089a9b06731"
```
Comment (debris) [inline Cargo.lock]: what is `patch.unused` property and how was it created? I've never seen this entry in `Cargo.lock` files

File: ethstore/src/json/key_file.rs [hunk 5]
```diff
@@ -160,7 +160,7 @@ impl<'a> Visitor<'a> for KeyFileVisitor {
 
 		let address = match address {
 			Some(address) => address,
-			None => return Err(V::Error::missing_field("address")),
+			None => H160::zero(),
```
Comment (debris) [inline ethstore/src/json/key_file.rs]: this is unacceptable and may lead to errors

**Comment by 5chdn** [comment]
Please rebase on master.

File: ethstore/src/account/crypto.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (tworec) [reply ethstore/src/account/crypto.rs]: bc `Pbkdf2` has `salt` as `Vec` now. But ok, I can pass `[u8]`  into [`crypto::derive_key_iterations`](https://github.com/paritytech/parity-common/blob/master/parity-crypto/src/lib.rs#L58)

File: ethstore/cli/src/main.rs (hunk 2 above)
Comment (tworec) [reply ethstore/cli/src/main.rs]: don't remember exactly, but i has some local issues with it. Restoring

File: Cargo.lock (hunk 3 above)
Comment (tworec) [reply Cargo.lock]: `ring` is problematic since it does not allow to link multiple versions of itself into program, and at Golem we are using `ring 0.13`

 https://github.com/briansmith/ring/blob/7e3682c02170f5eafa423b63bacfa018de2878a9/Cargo.toml#L14-L15

Comment (tworec) [reply Cargo.lock]: my final workaround is to extremely strip parity-ethereum project under [my private branch](https://github.com/tworec/parity-ethereum/tree/strip-down-ethkeys) to just contain `ethkey` and `ethstore` which we are using as for now.

Is there any other solution to have linked ring 12 and 13 in single program?

Comment (tworec) [reply Cargo.lock]: So for this PR I'm reverting this change

File: ethstore/src/json/key_file.rs (hunk 5 above)
Comment (tworec) [reply ethstore/src/json/key_file.rs]: ok, so I propose import with optional password to recover address when it is not in keystore
